vault/login_mfa.go

```diff
@@ -1725,6 +1730,10 @@ func (c *Core) validateLoginMFAInternal(ctx context.Context, methodID string, en
 			return fmt.Errorf("MFA secret for method name %q not present in entity %q", mConfig.Name, entity.ID)
 		}
 
+		if mfaFactors == nil {
```

Comment:
Ah, I see why you did it in the func previously, because it you did it right after the call to parseMfaFactors, you wouldn't know without the switch whether it was needed or not.

Now I'm a bit unsure how to proceed, because:
* if we added this check we'd presumably want it in `case mfaMethodTypeDuo` too, right?
* this check does nothing, as we never return a nil *MFAFactor from parseMfaFactors
* even when mfaFactors is non-nil, it may be empty, i.e. no creds were supplied, maybe parseMfaFactors should return nil in that case?

Reply:
In Duo, we already have check for that [here](https://github.com/hashicorp/vault/blob/0b0d26977fa547f8f617c3a13b47bbc17c4fb5ff/vault/login_mfa.go#L1886-L1889). Since passcode is not mandatory for Duo, I think it should be treated differently. 
For TOTP, we also check if `mfaFactors.passcode == ""` [here](https://github.com/hashicorp/vault/blob/0b0d26977fa547f8f617c3a13b47bbc17c4fb5ff/vault/login_mfa.go#L2339-L2341) and return an error for that. 
I agree that this check does nothing really. So, maybe it is better to just remove it instead of changing `parseMfaFactor` function to return nil in such a case. Because returning nil would fall into the same non-idiomatic Go anti-pattern 
```
func foo() (*thing, error) { return nil, nil }
```
Maybe I was too paranoid to think that `mfaFactors` could be nil somehow.

Reply:
Actually, thinking about possible future extensions of `parseMfaFactors`, that may introduce a `nil, nil` return case in the future. would it make sense to keep this logic that does nothing right now? 
My rational is that testing MFA is really hard and there are lot of combinations that could happen and can be easily missed. One way to go is to add a test for that, but then those tests could also be changed. Also, automating testing push notification is non-trivial. So, having such a check has no harm right now, and maybe even a good protection in the future? What do you think?

Reply:
I suggest you return nil instead of &mfaFactors{}; Go anti-pattern or no, `(nil, nil)` is widely used in Vault, so we're somewhat used to it, and thus it's a clearer signal of "I didn't find any creds" than an empty mfaFactors. 

Also be consistent and check for nil mfaFactors in the same place, either in the per-MFA validate funcs, or in the switch statement where we call them (I didn't realize validateDuo was already checking for nil mfaFactors when I made the initial comment). My preference would be in the switch statement, but it doesn't really matter.

Reply:
Sounds good. The only thing to consider for Duo, is that a `nil` mfaFactor is a valid case as Duo can be used with push notification. So, in the `ValidateDuo` function we check if the mfaFactor is not nil, and then make use of the `mfaFactor.passcode`. For that reason, and to be consistent, I will move the nil check inside validateTOTP as well.